### PR TITLE
Block-based ANN (BBAnn)

### PR DESCRIPTION
**Pull request by PwzXxm**
This is a submission for Track 2. We are a joint team from Zilliz and SUSTech.

### CODE REVIEW

**Comment by PwzXxm** [comment]
> > > @PwzXxm Here are the results for msspacev public query set for BBANN on Azure L8sv2 VM. Are they as expected?
> 
> > > bbann,BBANN,msspacev-1B,10,1610.4208408102975,0.0,1000000.0,11878352.0,7375.930377318827,0.0,0.0,0.7577705007504434 bbann,BBANN,msspacev-1B,10,1820.4236647258076,0.0,1000000.0,11878352.0,6525.048113890081,0.0,0.0,0.747574703233729 bbann,BBANN,msspacev-1B,10,1755.869157475597,0.0,1000000.0,11878352.0,6764.941424836825,0.0,0.0,0.7542058943921408 bbann,BBANN,msspacev-1B,10,1936.7360372815363,0.0,1000000.0,11878352.0,6133.1806561893845,0.0,0.0,0.7458589166325555 bbann,BBANN,msspacev-1B,10,3912.993335161897,0.0,1000000.0,11878352.0,3035.617743905138,0.0,0.0,0.6773809523809524 bbann,BBANN,msspacev-1B,10,1821.776089651613,0.0,1000000.0,11878352.0,6520.20413895736,0.0,0.0,0.7500818665575113 bbann,BBANN,msspacev-1B,10,1518.3505425395547,0.0,1000000.0,11878352.0,7823.1947545739795,0.0,0.0,0.7602026197298404 bbann,BBANN,msspacev-1B,10,1702.433482388304,0.0,1000000.0,11878352.0,6977.2781861269195,0.0,0.0,0.7522240414790559 bbann,BBANN,msspacev-1B,10,1629.0401932767472,0.0,1000000.0,11878352.0,7291.626105373854,0.0,0.0,0.7563787692727522 bbann,BBANN,msspacev-1B,10,2070.9883234427366,0.0,1000000.0,11878352.0,5735.595833903039,0.0,0.0,0.7408104789193615
> 
> > 
> 
> > These values match ours when ties are not considered in the ground truth list. We are working on it to improve the recall of this dataset.
> 
> > 
> 
> > Our index works better on range search, have you obtained any results for the SSNPP dataset?
> 
> 
> 
> bbann,BBANN,ssnpp-1B,60000,1656.4058759602651,0.0,1000000.0,41172124.0,24856.301584979203,0.0,0.0,0.47941143039246625
> 
> bbann,BBANN,ssnpp-1B,60000,1857.8207993202693,0.0,1000000.0,41172124.0,22161.51526297039,0.0,0.0,0.47366867391087714
> 
> bbann,BBANN,ssnpp-1B,60000,1849.7986209696735,0.0,1000000.0,41172124.0,22257.624983208912,0.0,0.0,0.47742031922137884
> 
> bbann,BBANN,ssnpp-1B,60000,1852.7778136811514,0.0,1000000.0,41172124.0,22221.83561136133,0.0,0.0,0.4755651927037486
> 
> bbann,BBANN,ssnpp-1B,60000,1857.8624555169647,0.0,1000000.0,41172124.0,22161.01836696169,0.0,0.0,0.471688981298477
> 
> bbann,BBANN,ssnpp-1B,60000,1782.1072035608154,0.0,1000000.0,41172124.0,23103.056829428824,0.0,0.0,0.4760617702502948
> 
> bbann,BBANN,ssnpp-1B,60000,1657.87953405736,0.0,1000000.0,41172124.0,24834.207283588745,0.0,0.0,0.47938859327509176
> 
> bbann,BBANN,ssnpp-1B,60000,2097.729515424244,0.0,1000000.0,41172124.0,19626.993707848636,0.0,0.0,0.4730428590410072
> 
> bbann,BBANN,ssnpp-1B,60000,1541.8264590489878,0.0,1000000.0,41172124.0,26703.474803121055,0.0,0.0,0.48058676339915957
> 
> bbann,BBANN,ssnpp-1B,60000,2310.3409142367714,0.0,1000000.0,41172124.0,17820.800275097645,0.0,0.0,0.47038895244066503
> 
> 
> 
> Does that align with what you see?

It seems like you are using radius 60000 instead of 96237 as you mentioned in the [issue #51](https://github.com/harsha-simhadri/big-ann-benchmarks/issues/51).

**Comment by PwzXxm** [comment]
> > > > @PwzXxm Here are the results for msspacev public query set for BBANN on Azure L8sv2 VM. Are they as expected?
> > 
> > > > bbann,BBANN,msspacev-1B,10,1610.4208408102975,0.0,1000000.0,11878352.0,7375.930377318827,0.0,0.0,0.7577705007504434 bbann,BBANN,msspacev-1B,10,1820.4236647258076,0.0,1000000.0,11878352.0,6525.048113890081,0.0,0.0,0.747574703233729 bbann,BBANN,msspacev-1B,10,1755.869157475597,0.0,1000000.0,11878352.0,6764.941424836825,0.0,0.0,0.7542058943921408 bbann,BBANN,msspacev-1B,10,1936.7360372815363,0.0,1000000.0,11878352.0,6133.1806561893845,0.0,0.0,0.7458589166325555 bbann,BBANN,msspacev-1B,10,3912.993335161897,0.0,1000000.0,11878352.0,3035.617743905138,0.0,0.0,0.6773809523809524 bbann,BBANN,msspacev-1B,10,1821.776089651613,0.0,1000000.0,11878352.0,6520.20413895736,0.0,0.0,0.7500818665575113 bbann,BBANN,msspacev-1B,10,1518.3505425395547,0.0,1000000.0,11878352.0,7823.1947545739795,0.0,0.0,0.7602026197298404 bbann,BBANN,msspacev-1B,10,1702.433482388304,0.0,1000000.0,11878352.0,6977.2781861269195,0.0,0.0,0.7522240414790559 bbann,BBANN,msspacev-1B,10,1629.0401932767472,0.0,1000000.0,11878352.0,7291.626105373854,0.0,0.0,0.7563787692727522 bbann,BBANN,msspacev-1B,10,2070.9883234427366,0.0,1000000.0,11878352.0,5735.595833903039,0.0,0.0,0.7408104789193615
> > 
> > > 
> > 
> > > These values match ours when ties are not considered in the ground truth list. We are working on it to improve the recall of this dataset.
> > 
> > > 
> > 
> > > Our index works better on range search, have you obtained any results for the SSNPP dataset?
> > 
> > 
> > 
> > bbann,BBANN,ssnpp-1B,60000,1656.4058759602651,0.0,1000000.0,41172124.0,24856.301584979203,0.0,0.0,0.47941143039246625
> > 
> > bbann,BBANN,ssnpp-1B,60000,1857.8207993202693,0.0,1000000.0,41172124.0,22161.51526297039,0.0,0.0,0.47366867391087714
> > 
> > bbann,BBANN,ssnpp-1B,60000,1849.7986209696735,0.0,1000000.0,41172124.0,22257.624983208912,0.0,0.0,0.47742031922137884
> > 
> > bbann,BBANN,ssnpp-1B,60000,1852.7778136811514,0.0,1000000.0,41172124.0,22221.83561136133,0.0,0.0,0.4755651927037486
> > 
> > bbann,BBANN,ssnpp-1B,60000,1857.8624555169647,0.0,1000000.0,41172124.0,22161.01836696169,0.0,0.0,0.471688981298477
> > 
> > bbann,BBANN,ssnpp-1B,60000,1782.1072035608154,0.0,1000000.0,41172124.0,23103.056829428824,0.0,0.0,0.4760617702502948
> > 
> > bbann,BBANN,ssnpp-1B,60000,1657.87953405736,0.0,1000000.0,41172124.0,24834.207283588745,0.0,0.0,0.47938859327509176
> > 
> > bbann,BBANN,ssnpp-1B,60000,2097.729515424244,0.0,1000000.0,41172124.0,19626.993707848636,0.0,0.0,0.4730428590410072
> > 
> > bbann,BBANN,ssnpp-1B,60000,1541.8264590489878,0.0,1000000.0,41172124.0,26703.474803121055,0.0,0.0,0.48058676339915957
> > 
> > bbann,BBANN,ssnpp-1B,60000,2310.3409142367714,0.0,1000000.0,41172124.0,17820.800275097645,0.0,0.0,0.47038895244066503
> > 
> > 
> > 
> > Does that align with what you see?
> 
> It seems like you are using radius 60000 instead of 96237 as you mentioned in the [issue #51](https://github.com/harsha-simhadri/big-ann-benchmarks/issues/51).
> 
> 

Also, we have calculated the distances among vectors in GT files and its corresponding query. There are distances more than 60000 and none of them are further than 96237. Thus, we reckon that the radius should be 96237 as you said before.

**Comment by PwzXxm** [comment]
Just wondering did you count ties when getting the above results for msspacev and ssnpp?

File: algos.yaml [hunk 1]
```diff
@@ -493,6 +548,38 @@ text2image-1B:
...
               {"Ls":100, "BW":10, "T":16}]
+    bbann: 
+      docker-tag: billion-scale-benchmark-bbann
+      module: benchmark.algorithms.bbann
+      constructor: BbANN
+      base-args: ["@metric"]
+      run-groups:
+        base:
+          args: |
+              [{"hnswM":32, "hnswefC":500, "K1":128, "blockSize":4096, "vector_use_sq":true, "use_hnsw_sq": true, "identifier":"bbann"}]
```
[the remark this answers is not in the document]
Comment (PwzXxm) [reply algos.yaml]: The index for text2image has been uploaded and the YAML file has been updated.

**Comment by harsha-simhadri** [comment]
> Just wondering did you count ties when getting the above results for msspacev and ssnpp?

Not so on the public query datasets.

**Comment by harsha-simhadri** [comment]
> > > > @PwzXxm Here are the results for msspacev public query set for BBANN on Azure L8sv2 VM. Are they as expected?
> > 
> > 
> > > > bbann,BBANN,msspacev-1B,10,1610.4208408102975,0.0,1000000.0,11878352.0,7375.930377318827,0.0,0.0,0.7577705007504434 bbann,BBANN,msspacev-1B,10,1820.4236647258076,0.0,1000000.0,11878352.0,6525.048113890081,0.0,0.0,0.747574703233729 bbann,BBANN,msspacev-1B,10,1755.869157475597,0.0,1000000.0,11878352.0,6764.941424836825,0.0,0.0,0.7542058943921408 bbann,BBANN,msspacev-1B,10,1936.7360372815363,0.0,1000000.0,11878352.0,6133.1806561893845,0.0,0.0,0.7458589166325555 bbann,BBANN,msspacev-1B,10,3912.993335161897,0.0,1000000.0,11878352.0,3035.617743905138,0.0,0.0,0.6773809523809524 bbann,BBANN,msspacev-1B,10,1821.776089651613,0.0,1000000.0,11878352.0,6520.20413895736,0.0,0.0,0.7500818665575113 bbann,BBANN,msspacev-1B,10,1518.3505425395547,0.0,1000000.0,11878352.0,7823.1947545739795,0.0,0.0,0.7602026197298404 bbann,BBANN,msspacev-1B,10,1702.433482388304,0.0,1000000.0,11878352.0,6977.2781861269195,0.0,0.0,0.7522240414790559 bbann,BBANN,msspacev-1B,10,1629.0401932767472,0.0,1000000.0,11878352.0,7291.626105373854,0.0,0.0,0.7563787692727522 bbann,BBANN,msspacev-1B,10,2070.9883234427366,0.0,1000000.0,11878352.0,5735.595833903039,0.0,0.0,0.7408104789193615
> > 
> > 
> > > 
> > 
> > 
> > > These values match ours when ties are not considered in the ground truth list. We are working on it to improve the recall of this dataset.
> > 
> > 
> > > 
> > 
> > 
> > > Our index works better on range search, have you obtained any results for the SSNPP dataset?
> > 
> > 
> > bbann,BBANN,ssnpp-1B,60000,1656.4058759602651,0.0,1000000.0,41172124.0,24856.301584979203,0.0,0.0,0.47941143039246625
> > bbann,BBANN,ssnpp-1B,60000,1857.8207993202693,0.0,1000000.0,41172124.0,22161.51526297039,0.0,0.0,0.47366867391087714
> > bbann,BBANN,ssnpp-1B,60000,1849.7986209696735,0.0,1000000.0,41172124.0,22257.624983208912,0.0,0.0,0.47742031922137884
> > bbann,BBANN,ssnpp-1B,60000,1852.7778136811514,0.0,1000000.0,41172124.0,22221.83561136133,0.0,0.0,0.4755651927037486
> > bbann,BBANN,ssnpp-1B,60000,1857.8624555169647,0.0,1000000.0,41172124.0,22161.01836696169,0.0,0.0,0.471688981298477
> > bbann,BBANN,ssnpp-1B,60000,1782.1072035608154,0.0,1000000.0,41172124.0,23103.056829428824,0.0,0.0,0.4760617702502948
> > bbann,BBANN,ssnpp-1B,60000,1657.87953405736,0.0,1000000.0,41172124.0,24834.207283588745,0.0,0.0,0.47938859327509176
> > bbann,BBANN,ssnpp-1B,60000,2097.729515424244,0.0,1000000.0,41172124.0,19626.993707848636,0.0,0.0,0.4730428590410072
> > bbann,BBANN,ssnpp-1B,60000,1541.8264590489878,0.0,1000000.0,41172124.0,26703.474803121055,0.0,0.0,0.48058676339915957
> > bbann,BBANN,ssnpp-1B,60000,2310.3409142367714,0.0,1000000.0,41172124.0,17820.800275097645,0.0,0.0,0.47038895244066503
> > Does that align with what you see?
> 
> It seems like you are using radius 60000 instead of 96237 as you mentioned in the [issue #51](https://github.com/harsha-simhadri/big-ann-benchmarks/issues/51).
> 
> Also, we have calculated the distances among vectors in GT files and its corresponding query. There are distances more than 60000 and none of them are further than 96237. Thus, we reckon that the radius should be 96237 as you said before.

With radius square 96237, I see

bbann,BBANN,ssnpp-1B,60000,2097.729515424244,0.0,1000000.0,41172124.0,19626.993707848636,0.0,0.0,0.4730428590410072
bbann,BBANN,ssnpp-1B,96237,1462.0503482699048,0.0,1000000.0,41172708.0,28160.937172048216,0.0,0.0,0.8863012589479978
bbann,BBANN,ssnpp-1B,96237,1671.0205269554601,0.0,1000000.0,41172708.0,24639.259264525735,0.0,0.0,0.8772537380208939
bbann,BBANN,ssnpp-1B,96237,1516.7019615432832,0.0,1000000.0,41172708.0,27146.208710711835,0.0,0.0,0.8857355212675846
bbann,BBANN,ssnpp-1B,96237,1587.3311664137136,0.0,1000000.0,41172708.0,25938.322683490336,0.0,0.0,0.8816703846185786
bbann,BBANN,ssnpp-1B,96237,1707.1922937796771,0.0,1000000.0,41172708.0,24117.2058648676,0.0,0.0,0.872364908070222
bbann,BBANN,ssnpp-1B,96237,1545.1303206752962,0.0,1000000.0,41172708.0,26646.75428931169,0.0,0.0,0.8824276426697005
bbann,BBANN,ssnpp-1B,96237,1455.0100602629188,0.0,1000000.0,41172708.0,28297.19815996333,0.0,0.0,0.8861660528326329
bbann,BBANN,ssnpp-1B,96237,1767.7956237172189,0.0,1000000.0,41172708.0,23290.42308263237,0.0,0.0,0.8772940922453454
bbann,BBANN,ssnpp-1B,96237,1380.1362515617418,0.0,1000000.0,41172708.0,29832.350214270202,0.0,0.0,0.8889190932833961
bbann,BBANN,ssnpp-1B,96237,1905.0949488596978,0.0,1000000.0,41172708.0,21611.892900479364,0.0,0.0,0.8731373476141441

**Comment by PwzXxm** [comment]
> @PwzXxm Based on the results I pasted above on this thread, and applying the 1500QPS cut off, I see the following results for the 3 datasets you submitted (Without ties in msspacev):
> 
> msspacev-1B: 0.7602 ssnpp-1B: 0.88573 text2image-1B: 0.487202

For text2image-1B, I saw the following entry:

> bbann,BBANN,text2image-1B,10,1540.622672933968,0.0,1000000.0,29064836.0,18865.64212679592,0.0,0.0,0.495423

Is the fifth column the QPS? If yes, then the recall is 0.495423, right?

**Comment by harsha-simhadri** [comment]
Could you please rebase from main and squash your commits into 1 so its cleaner to rebase and merge?

**Comment by PwzXxm** [comment]
> Could you please rebase from main and squash your commits into 1 so its cleaner to rebase and merge?
Done. Thank you for your patience.

**Review by harsha-simhadri — APPROVED** [review]

**Pull request merged**